### PR TITLE
Key generation, typedefs and bug-fixes implemented. First Hello world example.

### PR DESCRIPTION
**Pull request by reicheratwork**
Fixed bug in ostream with varargs list being re-used incorrectly.
Fixed bug in ostream with buffers being transferred incorrectly.
Added key calculation functions to the streamer generator.
Added typedef streaming function generation.
Added unittests for key calculation and typedefs.
Moved the helloworld serdata and sertopic classes to templates to make it easier to create topics using the generated classes.

### CODE REVIEW

**Comment by k0ekk0ek** [comment]
@reicheratwork, I changed your PR to go into `idlcxx` instead of `master`, which is probably a better idea? Unless, of course, it's ready to go there instead, in which case I'd be delighted to change it back :wink: :stuck_out_tongue_winking_eye:

File: examples/helloworld/HelloWorldData.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    userID(*((int32_t*)((char*)data + position)));  //reading bytes for member: userID
+    position += 4;  //moving position indicator
+    uint32_t sequenceentries = *((uint32_t*)((char*)data + position));  //number of entries in the sequence
+    position += 4;  //moving position indicator
+    message().assign((char*)((char*)data + position), (char*)((char*)data + position) + sequenceentries);  //putting data into container
+    position += sequenceentries * 1;  //moving position indicator
+    return position;
+  }
+
+}
```
Comment (k0ekk0ek) [inline examples/helloworld/HelloWorldData.cpp:42]: This probably shouldn't be here? Same goes for the other generated files.

File: examples/helloworld/HelloWorldData.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+#ifdef OSPL_DDS_CXX11
+        void message(std::string&& _val_) { this->message_ = _val_; }
+#endif
+        size_t write_struct(void* data, size_t position) const;
+        size_t write_size(size_t offset) const;
+        size_t read_struct(const void* data, size_t position);
+    };
+}
+
+#endif /* _HELLOWORLDDATA_H_ */
```
Comment (k0ekk0ek) [inline examples/helloworld/HelloWorldData.h:78]: Please remove this as well.

File: examples/helloworld/HelloWorldData_DCPS.hpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+{
+    static std::string value()
+    {
+        return org::eclipse::cyclonedds::topic::TopicTraits<HelloWorldData::Msg>::getTypeName();
+    }
+};
+}}
+
+REGISTER_TOPIC_TYPE(HelloWorldData::Msg)
+#endif /* ISOCPP_HELLOWORLDDATA_H */
```
Comment (k0ekk0ek) [inline examples/helloworld/HelloWorldData_DCPS.hpp:79]: Please remove.

File: examples/helloworld/publisher.cpp [hunk 4]
```diff
@@ -31,7 +31,7 @@ int main() {
         dds::domain::DomainParticipant participant(domain::default_id());
 
         /* To publish something, a topic is needed. */
-        dds::topic::Topic<HelloWorldData::Msg> topic(participant, "ddscxx_helloworld_example");
+        dds::topic::Topic<HelloWorldData::Msg> topic(participant, "HelloWorldData_Msg");
```
Comment (k0ekk0ek) [inline examples/helloworld/publisher.cpp:34]: I changed this one so we could test with a `C` reader/writer. I think it makes sense to change it, but maybe @eboasson has a different opinion on the matter?

File: src/ddscxx/include/dds/topic/detail/TTopicImpl.hpp [hunk 5]
```diff
@@ -187,7 +185,9 @@ dds::topic::detail::Topic<T>::Topic(const dds::domain::DomainParticipant& dp,
 
     this->listener(listener, mask);
 
+#if 0
     this->AnyTopicDelegate::set_copy_out(org::eclipse::cyclonedds::topic::TopicTraits<T>::getCopyOut());
+#endif
```
Comment (k0ekk0ek) [inline src/ddscxx/include/dds/topic/detail/TTopicImpl.hpp:190]: Probably better to just remove these lines?

**Review by k0ekk0ek — COMMENTED** [review]
Thanks @reicheratwork! Had a first look. There's probably some remnants from testing here and there? I'll continue this tomorrow.

File: src/ddscxx/include/org/eclipse/cyclonedds/sub/AnyDataReaderDelegate.hpp [hunk 6]
```diff
@@ -102,6 +102,14 @@ class OMG_DDS_API AnyDataReaderDelegate : public org::eclipse::cyclonedds::core:
...
     virtual ~AnyDataReaderDelegate();
 
+
+    virtual void copy_samples(
+      dds::sub::detail::SamplesHolder& samples,
+      void**& c_sample_pointers,
+      dds_sample_info_t*& c_sample_infos,
+      int num_read) {
+      assert(1);
+    }
```
Comment (k0ekk0ek) [inline src/ddscxx/include/org/eclipse/cyclonedds/sub/AnyDataReaderDelegate.hpp:112]: Is it your intention to provide the implementation in the header file? I'm not too familiar with the C++ api, but this seems like an unnecessary change? @e-hndrks, can you shed some light perhaps?

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+    auto d = new ddscxx_serdata<T>(topic, kind);
+
+    if (kind != SDK_DATA) {
+      //???
+    }
+    else /*if (!topic->is_request_header)*/ {
+      auto msg = static_cast<const T*>(sample);
+
+      size_t sz = msg->write_size(0);
+      //fprintf(stderr, "sz is: %zu\n", sz);
```
Comment (k0ekk0ek) [inline src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:178]: Please remove

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+        *(ptr + 1) = 0x1;
+
+      msg->write_struct(calc_offset(d->data(), 4), 0);
+
+      /*fprintf(stderr, "raw message contents:\n");
+      for (size_t s = 0; s < sz + 4; s++)
+      {
+        fprintf(stderr, "%02x ", ptr[s]);
+      }
+      fprintf(stderr, "\n");*/
```
Comment (k0ekk0ek) [inline src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:192]: Please remove

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+  size_t size() const { return m_size; }
+  void* data() const { return m_data.get(); }
+};
+
+template <typename T>
+bool serdata_eqkey(const struct ddsi_serdata* a, const struct ddsi_serdata* b)
+{
+  (void)a;
+  (void)b;
+  return true;
```
Comment (k0ekk0ek) [inline src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:71]: Don't know how much of this is actually required, but it kind of seems that keys aren't compared and just always match? Am I correct in my assumption?

File: src/idlcxx/tests/streamer_generator.c [hunk 10]
```diff
@@ -19,6 +19,23 @@
...
 #include "CUnit/Theory.h"
 
+#define visual_string_cmp 1
+
+#define test_string_diff3(value_tested,value_tested_for) \
+{ \
+  if (strcmp(value_tested,value_tested_for)) { \
+    if (visual_string_cmp) { \
+      FILE* fa = fopen("/tmp/generated.txt","w"); \
+      FILE * fb = fopen("/tmp/tested.txt","w"); \
```
Comment (k0ekk0ek) [inline src/idlcxx/tests/streamer_generator.c:29]: This probably shouldn't be here?

**Review by k0ekk0ek — CHANGES_REQUESTED** [review]
@reicheratwork, I should probably actually test it, but it seems there's some polishing required, and I think you should update the HelloWorld example to actually use the compiler.

File: examples/helloworld/HelloWorldData.cpp (hunk 1 above)
Comment (reicheratwork) [reply examples/helloworld/HelloWorldData.cpp:42]: The files helloworlddata.cpp and helloworlddata.h were placed here as a placeholder for the code that is supposed to be generated by idlc + idlcxx.
My personal opinion is to only remove these files when the code generation (idlcxx) is implemented completely.

File: examples/helloworld/HelloWorldData.h (hunk 2 above)
Comment (reicheratwork) [reply examples/helloworld/HelloWorldData.h:78]: Same comment as helloworddata.cpp.

File: examples/helloworld/HelloWorldData_DCPS.hpp (hunk 3 above)
Comment (reicheratwork) [reply examples/helloworld/HelloWorldData_DCPS.hpp:79]: Same comment as helloworlddata.cpp

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp (hunk 9 above)
Comment (reicheratwork) [reply src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:71]: At the moment this code was written, the key generation was not yet in the generated code, it now is.
I will replace these with the correct key calculation functions.

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp (hunk 8 above)
Comment (reicheratwork) [reply src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:192]: Will do.

File: src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp (hunk 7 above)
Comment (reicheratwork) [reply src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp:178]: Will do.

File: src/idlcxx/tests/streamer_generator.c (hunk 10 above)
Comment (reicheratwork) [reply src/idlcxx/tests/streamer_generator.c:29]: You are correct, this is a remnant from when I was testing the code.

File: src/ddscxx/include/dds/topic/detail/TTopicImpl.hpp (hunk 5 above)
Comment (reicheratwork) [reply src/ddscxx/include/dds/topic/detail/TTopicImpl.hpp:190]: I will look into this with @e-hndrks .

Comment (reicheratwork) [reply src/ddscxx/include/dds/topic/detail/TTopicImpl.hpp:190]: Will do.

File: examples/helloworld/CMakeLists.txt [hunk 11]
```diff
@@ -12,15 +12,15 @@
...
 
-# Convenience function, provided by the Idlpp-cxx that generates a CMake
-# target for the given IDL file. The function calls Idlcpp-cxx to generate
-# source files and compiles them into a library.
-idl_ddscxx_generate(ddscxxHelloWorldData_lib "HelloWorldData.idl")
+set(sources HelloWorldData.cpp)
+set(headers HelloWorldData.h HelloWorldData_DCPS.hpp)
+
+add_library(ddscxxHelloWorldData_lib OBJECT ${sources} ${headers})
+target_link_libraries(ddscxxHelloWorldData_lib PUBLIC  CycloneDDS-CXX::ddscxx)
```
[the remark this answers is not in the document]
Comment (reicheratwork) [reply examples/helloworld/CMakeLists.txt:23]: I will remove this.

File: src/ddscxx/include/org/eclipse/cyclonedds/sub/AnyDataReaderDelegate.hpp (hunk 6 above)
Comment (reicheratwork) [reply src/ddscxx/include/org/eclipse/cyclonedds/sub/AnyDataReaderDelegate.hpp:112]: This function needed to be implemented to get the helloworld example with the templates to compile.

File: examples/helloworld/publisher.cpp (hunk 4 above)
Comment (reicheratwork) [reply examples/helloworld/publisher.cpp:34]: I will change this back.

**Comment by k0ekk0ek** [comment]
Thanks @reicheratwork! I've included your work in #17.
